sound/soc/sof/ipc4-topology.c

```diff
@@ -1399,6 +1474,80 @@ static int sof_ipc4_prepare_src_module(struct snd_sof_widget *swidget,
 	return 0;
 }
 
+static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
+					   struct snd_pcm_hw_params *fe_params,
+					   struct snd_sof_platform_stream_params *platform_params,
+					   struct snd_pcm_hw_params *pipeline_params, int dir)
+{
+	struct snd_soc_component *scomp = swidget->scomp;
+	struct snd_soc_dapm_widget *widget = swidget->widget;
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct sof_ipc4_process *process = swidget->private;
+	struct snd_sof_control *scontrol = NULL;
+	void *extra_flag;
+	const struct snd_kcontrol_new *kc;
+	struct soc_bytes_ext *sbe;
+	struct sof_ipc4_control_data *control_data;
+	void *cfg = process->ipc_config_data;
+	int ret, i;
+
+	process->available_fmt.ref_audio_fmt = &process->available_fmt.base_config->audio_fmt;
+
+	/*
+	 * Output format is optional for process modules.
+	 * Process modules should handle the output format itself.
+	 */
+	ret = sof_ipc4_init_audio_fmt(sdev, swidget, &process->base_config,
+				      NULL, pipeline_params, &process->available_fmt,
+				      sizeof(process->base_config));
+	if (ret < 0)
+		return ret;
+
+	/* update pipeline memory usage */
+	sof_ipc4_update_pipeline_mem_usage(sdev, swidget, &process->base_config);
+
+	/*
+	 * 1. copy base_config to ipc_config_data
+	 * 2. copy output fmts to ipc_config_data
+	 * 3. copy extra payload to ipc_config_data
+	 */
+	/* update base_config */
+	memcpy(cfg, &process->base_config, sizeof(process->base_config));
+	cfg = cfg + sizeof(process->base_config);
+
+	/* update output fmt & update output hw_params */
+	for (i = 0; i < swidget->output_cnt; i++) {
+		/* TODO: to handle different output formats */
+		memcpy(cfg, process->available_fmt.out_audio_fmt,
+		       sizeof(struct sof_ipc4_audio_format));
+		cfg = cfg + sizeof(struct sof_ipc4_audio_format);
+	}
+
+	/* update extra payload */
+	for (i = 0; i < widget->num_kcontrols; i++) {
+		kc = &widget->kcontrol_news[i];
+		if (!kc)
+			return -ENODEV;
+
+		/* payload uses byte kcontrol */
+		if (widget->dobj.widget.kcontrol_type[i] != SND_SOC_TPLG_TYPE_BYTES)
+			continue;
+		sbe = (struct soc_bytes_ext *)kc->private_value;
+		scontrol = sbe->dobj.private;
+		if (!scontrol)
+			continue;
+		if (!sof_ipc4_ctl_for_inst_init(scontrol))
+			continue;
+
+		control_data = scontrol->ipc_control_data;
+		extra_flag = (void *)control_data->data->data;
+		memcpy(cfg, extra_flag, control_data->data->size);
+		break;
+	}
```

Comment:
I dont get this part. This is already done when you parse the byte controls for each widget, why parse again?

Reply:
This code is mainly to copy the kcontrol blob data to ipc4 message payload.

Reply:
@libinyang this is duplicating the work the common topology code has already done. In your 3rd patch you added support for copying the control data into control_data->data. Why not simply take the data from the scontrol->control_data?

Reply:
You look for the first kcontrol wich matches SND_SOC_TPLG_TYPE_BYTES and SOF_IPC4_MOD_INIT_INSTANCE then break out and skip the rest of the controls?
Can a module have multiple BYTES control that is INIT? (assuming a no answer)